### DISCUSSION TITLE
ACID Compliance for Milvus 2.0

### DISCUSSION THREAD

**Discussion by michaelteo** [milvus-io/milvus#16651] [Q&A and General discussion] [answered]
We are trying to assess Milvus 2.0 as a database. One of the key questions was around ACID compliance. based on what I understand Consistency is something that Milvus does not comply to by default because of Bounded Consistency. To be able to meet Consistency of ACID - i need to configure Consistency to Strong Consistency. Is this correct understanding.

Also I tried to summarize understanding of how Milvus can be ACID compliant. Are these points correct?

**Atomicity:**
The requests’ execution order is strictly in accordance with their issue-time order.

> For a batch request, all its sub-requests share the same issue time. In cases there are multiple proxies, issue time from different proxies are regarded as coming from a central clock. A batch insert/delete is guaranteed to become visible atomically.

Reference: http://github.com/milvus-io/milvus/blob/master/docs/developer_guides/chap01_system_overview.md#11-data-model

Also Milvus implements WAL that is should to guarantee Atomicity. 
Reference: https://www.cs.purdue.edu/homes/csjgwang/pubs/SIGMOD21_Milvus.pdf

**Consistency:**

Milvus supports 4 levels of consistency that can be configured based on use case - Link: https://milvus.io/blog/2022-1-27-milvus-2-0-a-glimpse-at-new-features.md#Consistency-level

Consistency can be set to strong to force to meet the consistency requirements for ACID if needed

However – vector database use cases usually have a tolerance for eventual consistency where performance is valued over some consistency. Parameter exists to guarantee consistency within a tolerable range

**Isolation:**

Based on my understanding - in Milvus maintain a global timestamp for all data insertion requests. This should guarantee there is a specific order for all transactions. To achieve true isolation, I believe we also need to set the Consistency Level to Strong.

Link: https://milvus.io/blog/deep-dive-4-data-insertion-and-data-persistence.md#Root-coord-and-Time-Tick

**Durability:**

Based on my understanding Milvus Durability guarantee is not there by default, but has to be forced programmatically if required.

> “Milvus returns success when inserted data is loaded to the message queue. However, the data is not yet flushed to the disk. Then Milvus’ data node writes the data in the message queue to persistent storage as incremental logs. If flush() is called, the data node is forced to write all data in the message queue to persistent storage immediately.”

References:
https://milvus.io/docs/product_faq.md#How-does-Milvus-flush-data
https://milvus.io/blog/deep-dive-4-data-insertion-and-data-persistence.md

**Comment by xiaofan-luan** [comment, marked as answer]
Hey Michael, thanks for the very detailed question.  
For milvus, since we are not only a database, but also a AI infra, the consistency is not as much important as a Transactional database.
If you'd like to, milvus support durability by default, support consistency if set search to strong(but should really suffer from the bad search performance), and isolation can be handled since search has global tso, which is also known as MVCC.

The trick here is Atomicity, if your do batch insertion, we don't guarantee all the batch will succeed in one run. It might happen that milvus return error but some of the data is actully written in

**Reply by xiaofan-luan** [reply]
I highly recommend you not to use strong consistency, while design Milvus we take performance/availability far seriously than consistency and all transactional stuffs like a OLTP database

**Reply by michaelteo** [reply]
Thanks - I agree that Milvus is not just a database and not primarily an OLTP database. These were the typical questions asked by our enterprise architect on databases. While they are familiar with non-OLTP distributed database with eventual consistency, these are the first questions to ask to make sure we understand the technology

Maybe the important idea you raise: the question of ACID compliance might be not relevant for Milvus use cases where high performance search/query is the main driver vs being a transactional system. 

And if ACID compliance matters a lot - maybe the better architecture is to have a separate OLTP system to keep initial data before loading into Milvus so there are ACID Guarantees before being loaded to Milvus as a AI Search Database. 

In this way, this is more similar to something like elasticsearch which should not be the primary data store but is used for fast document search.

**Reply by michaelteo** [reply]
And to clarify for Atomicity Point - will it eventually be consistent for batch insert even if milvus throws an error ?

Or it would become an issue ? Because if milvus recovers/restarts, it will try to read from the log and rewrite the whole batch insert again. But the partial insert is written already and milvus does not have record of this partial insert state due to error. 

So if Atomicity it truly required, then batch insert should be avoided, but load item by item/ "row-wise"
